torch_np/_wrapper.py

```diff
@@ -258,13 +268,60 @@ def corrcoef(x, y=None, rowvar=True, bias=NoValue, ddof=NoValue, *, dtype=None):
 
 
 def concatenate(ar_tuple, axis=0, out=None, dtype=None, casting="same_kind"):
-    if casting != "same_kind":
-        raise NotImplementedError
-    tensors = tuple(asarray(ar).get() for ar in ar_tuple)
-    if dtype is not None:
-        # XXX: map numpy dtypes
-        tensors = tuple(ar.type(dtype) for ar in ar_typle)
-    return asarray(torch.cat(tensors, axis, out=out))
+    if out is not None:
+        if dtype is not None:
+            # mimic numpy
+            raise TypeError("concatenate() only takes `out` or `dtype` as an "
+                            "argument, but both were provided.")
+        if not isinstance(out, ndarray):
+            raise ValueError("'out' must be an array")
+        out_tensor = out.get()
+    else:
+        out_tensor = out
```

Comment:
Fantastic candidate to be factored out into a function.

Reply:
Indeed, if only the logic generalizes. E.g. unary ufuncs seem to accept both `dtype=` and `out=`. (Not that I understand what `np.sin(x, dtype=..., out=...)` does, but it definitely accepts both)

Reply:
That's fine, right? If both are given, then `dtype == `out.dtype` should be true, otherwise you raise. And if it's true, there is no issue. I'd expect that to generalize to all functions.

Reply:
Kind of.

```
In [550]: np.add(np.ones(3), np.ones(3), dtype=float, out=np.empty(3))     # ufuncs accept both
Out[550]: array([2., 2., 2.])

In [551]: np.concatenate((np.ones(3), np.ones(4)), dtype=float, out=np.empty(7))    # concatenates does not
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [551], in <cell line: 1>()
----> 1 np.concatenate((np.ones(3), np.ones(4)), dtype=float, out=np.empty(7))

File <__array_function__ internals>:180, in concatenate(*args, **kwargs)

TypeError: concatenate() only takes `out` or `dtype` as an argument, but both were provided.

In [554]: np.abs(np.ones(3), dtype='int32', out=np.empty(3, dtype='complex'))      # what is this doing?
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
Input In [554], in <cell line: 1>()
----> 1 np.abs(np.ones(3), dtype='int32', out=np.empty(3, dtype='complex'))

UFuncTypeError: Cannot cast ufunc 'absolute' input from dtype('float64') to dtype('int32') with casting rule 'same_kind'

In [555]: np.abs(np.ones(3), dtype='float32', out=np.empty(3, dtype='complex'))    # but not always
Out[555]: array([1.+0.j, 1.+0.j, 1.+0.j])
```

what `dtype=` does for ufuncs is a bit of a mystery to me (how does this interact with `.types` of registered loops), but it definitely accepts dtype != out.dtype. Maybe safe to side-step as long as actual computations are offloaded to pytorch and it does it own casting dance internally. 

So the suggestion for ufuncs is to ignore the `dtype` if `out is not None`. Or maybe ignore it completely even.

Reply:
This looks like a bug, or at least questionable behaviour, in numpy:
```python
>>> x = np.ones(3)  # has dtype float64
>>> np.add(x, x, dtype=np.float64, out=np.empty(3, dtype=np.float16))
array([2., 2., 2.], dtype=float16)
```
I can't find an existing open issue (but it could exist, this is hard to search for), nor a note in the docs on this. I think what is happening is that the semantics of `dtype` is not what you'd expect. Based on common usage experience and common sense, you'd want it to be the _output dtype_. It usually is, but that's only a consequence of its actual semantics: it selects the _dtype of the inner loops_ of the implementation underlying the function. 

For non-ufuncs, the `dtype` keyword is typically documented as determining the output dtype. And for ufuncs, that's effectively how it's used. The "assume users know numpy internals, and let them select inner loop casts" makes very little sense.

> So the suggestion for ufuncs is to ignore the `dtype` if `out is not None`. Or maybe ignore it completely even.

I think this is exactly an example like @lezcano suggested. Whatever we do, let's use a consistent design rather than follow the quirks of numpy exactly. Imho best to not have any distinction between what ufuncs and non-ufunc functions do here.

Reply:
I can guess a use case for the accumulator dtype being wider than the input dtype, as in `np.add(x, x, dtype=float64)` with `x.dtype==np.float16`. If we were designing it from scrach, I'd go for overloading the `out` argument into either an array to place the result into, or a dtype for a fresh array. But since we are not :-), and given that pytorch does not support `torch.add(x, x, dtype=....)` anyway, the two options I see are either raise or ignore. Raising seems less numpy-compatible, can make porting from numpy harder.

Reply:
IOW, for numpy itself, `dtype=...` might as well belong to "other obscure keyword args", somewhere together with `extobj`, https://numpy.org/doc/stable/reference/ufuncs.html#ufuncs-kwargs. Did somebody start planning numpy 2.0? :-).

Reply:
Yes, the accumulator thing is probably a/the reason for why it was added way back when. But that's extremely niche, and in the few cases where you actually need it, it should be done by default. 

> Did somebody start planning numpy 2.0? :-).

Yes, actually - a NEP should arrive shortly. 1.26 -> 2.0 (Dec'23)

Reply:
In PyTorch we have a `dtype=` kwarg for reductions. This makes complete sense. Now, the benefits of having it for a non-reduction operator I think that are fairly niche and better dropped out.

You can have a look at the semantics of `dtype=` (and of `out=` and other kwargs) within the context of a reduction in the generic implementation of a reduction in PrimTorch:
https://github.com/pytorch/pytorch/blob/fd3a7264ae80332ba5ec8f60446e6d7a2c2276c1/torch/_refs/__init__.py#L1992-L2003